src/Browser.php

```diff
@@ -335,4 +338,33 @@ public function withOptions(array $options)
 
         return $browser;
     }
+
+    /**
+    * Returns the CookieJar handler.
+    * @return CookieJar
+    */
+    public function getCookieJar()
+    {
+        return $this->cookieJar;
+    }
+
+    /**
+    * Set the CookieJar handler.
+    * This object is responsible for managing cookies in all requests.
+    * @param CookieJar $cookieJar
+    */
+    public function setCookieJar($cookieJar) {
+        $this->cookieJar = $cookieJar;
+
+        // to work with PHP 5.3
+        $jar = &$this->cookieJar;
+        $requestCallback = function($request) use ($jar) {
+            return $jar === null ? $request : $jar->onRequest($request);
+        };
+        $responseCallback = function($response, $request) use ($jar) {
+            return $jar === null ? $response : $jar->onResponse($response, $request);
+        };
+        $this->transaction->addRequestHandler($requestCallback, 'browser-cookie-jar');
+        $this->transaction->addResponseHandler($responseCallback, 'browser-cookie-jar');
```

Comment:
Interesting concept! (see below)

The current implementation seems to imply this only keeps adding to this list of handlers and never cleans this up after assigning different cookie jars and/or unsetting? It's my understanding that this API might be overcomplicating things here as we have no use cases beyond cookies for this at the moment. Addressing this would likely also solve your reference problem mentioned above.